src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2189,6 +2189,26 @@ void QgsVertexTool::moveVertex( const QgsPointXY &mapPoint, const QgsPointLocato
 
   QgsAbstractGeometry *geomTmp = geom.constGet()->clone();
 
+  // If moving point is not 3D but destination yes, check if it can be promoted
+  if ( layerPoint.is3D() && !geom.vertexAt( dragVertexId ).is3D() && QgsWkbTypes::hasZ( dragLayer->wkbType() ) )
```

Comment:
@domi4484 , why not simply checking geom.is3D(), it'll return the same value but will likely be less costly than retrieving a vertex and checking presence of a Z. AFAIK, you can't have mixed 3D and non-3D vertices in a 3D geometry.

Reply:
Oh thanks, when I started the implementation I was not sure that geometries can't be mixed